fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/ItemOverlayRenderer.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.rendereregistry.v1;
+
+import com.mojang.blaze3d.systems.RenderSystem;
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.font.TextRenderer;
+import net.minecraft.client.network.ClientPlayerEntity;
+import net.minecraft.client.render.BufferBuilder;
+import net.minecraft.client.render.Tessellator;
+import net.minecraft.client.render.VertexConsumerProvider;
+import net.minecraft.client.render.VertexFormats;
+import net.minecraft.client.util.math.MatrixStack;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.util.math.Matrix4f;
+import org.lwjgl.opengl.GL11;
+
+public interface ItemOverlayRenderer {
+	/**
+	 * @return <code>true</code> to cancel vanilla overlay rendering.
+	 */
+	boolean renderOverlay(MatrixStack matrixStack, float zOffset, TextRenderer renderer, ItemStack stack, int x, int y,
+						  String countLabel);
+
+	default void renderCountLabel(MatrixStack matrixStack, float zOffset, TextRenderer renderer, ItemStack stack,
+								  int x, int y, String countLabel) {
+		if (stack.getCount() != 1 || countLabel != null) {
+			String string = countLabel == null ? String.valueOf(stack.getCount()) : countLabel;
+			matrixStack.push();
+			matrixStack.translate(0.0D, 0.0D, zOffset + 200.0F);
+			VertexConsumerProvider.Immediate immediate = VertexConsumerProvider.immediate(Tessellator.getInstance().getBuffer());
+			renderer.draw(string, (float)(x + 19 - 2 - renderer.getWidth(string)), (float)(y + 6 + 3), 0xFFFFFF,
+			    true, matrixStack.peek().getModel(), immediate, false, 0, 0xF000F0);
+			immediate.draw();
+			matrixStack.pop();
+		}
+	}
+
+	@SuppressWarnings("deprecation")
+	default void renderDurabilityBar(MatrixStack matrixStack, float zOffset, ItemStack stack, int x, int y) {
```

Comment:
I dont think we really need to allow mods to control the whole rendering of the durablity bar. All that is needed is a way for a mod to provide a stack based durablity along with possibly the color of the durablity bar.

In reborn core I have this mixin https://github.com/TechReborn/RebornCore/blob/1.16/src/main/java/reborncore/mixin/client/MixinItemRenderer.java 

One idea for this could be to add a FabricItemSettings that allows this, or just keep it as an interface you put on the item.

This PR could possibly be a more generic way for a mod to render anything they want ontop of an item.

Reply:
large swathes of vanilla code have been yeeted
also yes, this is pretty generic